rust/vectordb/src/table/dataset.rs

```diff
@@ -0,0 +1,234 @@
+// Copyright 2024 LanceDB Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    ops::{Deref, DerefMut},
+    sync::Arc,
+    time::{self, Duration, Instant},
+};
+
+use lance::Dataset;
+use tokio::sync::{RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+use crate::error::Result;
+
+/// A wrapper around a [Dataset] that provides lazy-loading and consistency checks.
+///
+/// This can be cloned cheaply. It supports concurrent reads or exclusive writes.
+#[derive(Debug, Clone)]
+pub struct DatasetConsistencyWrapper(Arc<RwLock<DatasetRef>>);
+
+/// A wrapper around a [Dataset] that provides consistency checks.
+///
+/// The dataset is lazily loaded, and starts off as None. On the first access,
+/// the dataset is loaded.
```

Comment:
This struct is not public (not even `pub(crate)`).  Don't document or use `//`.

Reply:
Are you worried that this will show up in the reference docs?

Or do you think it's confusing when private APIs have documentation?

This entire file is private. Anything labelled `pub` here is only exposed to the crate, as the module is `pub(crate)`. Would it be sufficient to state that up front at the top of the file?

Reply:
> Or do you think it's confusing when private APIs have documentation?

Mostly this, and confusing might not be the right word.  Maybe it's just my personal preference too.  I try to be more sparse when documenting things that are internal but I don't know if that's a good thing or not.

Reply:
Let's talk about this more later, and we can decide some norms we want. For now, I'll merge this as-is.